File: util/stage/CastleList.java
package common.util.stage;

import common.io.json.JsonClass;
import common.io.json.JsonClass.RType;
import common.pack.FixIndexList.FixIndexMap;
import common.pack.IndexContainer;
import common.pack.PackData.UserPack;
import common.pack.UserProfile;
import common.system.VImg;
import common.system.files.VFile;
import common.util.stage.MapColc.PackMapColc;

import java.util.*;

@JsonClass
public abstract class CastleList extends FixIndexMap<CastleImg> implements IndexContainer {

	public static class DefCasList extends CastleList {

		public final String str;
		public final String id;

		public DefCasList(String hash, String name) {
			id = hash;
			str = name;
			map().put(id, this);
			defset().add(this);
			for (VFile vf : VFile.get("./org/img/" + name).list())
				if (!vf.getName().matches("ec[0-9]+\\.png"))
					add(new CastleImg(getNextID(CastleImg.class), new VImg(vf)));

		}

		@Override
		public String getSID() {
			return id;
		}

		@Override
		public String toString() {
			return str + " (" + size() + ")";
		}
	}

	@JsonClass(read = RType.FILL)
	public static class PackCasList extends CastleList {

		public final UserPack pack;

		public PackCasList(UserPack p) {
			pack = p;
			map().put(pack.desc.id, this);
		}

		@Override
		public String getSID() {
			return pack.getSID();
		}

		@Override
		public String toString() {
			return pack.desc.name + " (" + size() + ")";
		}
	}

	private static final String REG_CASTLE = "castle";

	private static final String REG_DEF_CASTLE = "def_castle";

	public static Set<CastleList> defset() {
		return UserProfile.getPool(REG_DEF_CASTLE, CastleList.class);
	}

	public static Collection<CastleList> from(Stage sta) {
		MapColc mc = sta.getCont().getCont();
		if (!(mc instanceof PackMapColc))
			return defset();
		List<CastleList> list = new ArrayList<>();
		list.addAll(defset());
		UserPack pack = ((PackMapColc) mc).pack;
		list.add(pack.castles);
		for (String rel : pack.desc.dependency)
			list.add(UserProfile.getUserPack(rel).castles);
		return list;
	}

	@ContGetter
	public static CastleList getList(String str) {
		return map().get(str);
	}

	public static Map<String, CastleList> map() {
		return UserProfile.getRegister(REG_CASTLE, CastleList.class);
	}

	private CastleList() {
		super(CastleImg.class);
	}

	@SuppressWarnings("rawtypes")
	@Override
	public <R> R getList(Class cls, Reductor<R, FixIndexMap> func, R def) {
		return func.reduce(def, this);
	}

}

File: util/stage/CastleList.kt
package common.util.stage

import common.pack.FixIndexList.FixIndexMap
import common.pack.IndexContainer
import common.pack.IndexContainer.ContGetter
import common.pack.IndexContainer.Reductor
import common.pack.PackData.UserPack
import common.pack.UserProfile
import common.system.VImg
import common.system.files.VFile
import common.util.stage.CastleList
import common.util.stage.MapColc.PackMapColc
import java.util.*

abstract class CastleList private constructor() : FixIndexMap<CastleImg?>(CastleImg::class.java), IndexContainer {
    class DefCasList(val id: String, val str: String) : CastleList() {
        override fun getID(): String? {
            return id
        }

        override fun toString(): String {
            return str + " (" + size() + ")"
        }

        init {
            map()[id] = this
            defset().add(this)
            for (vf in VFile.Companion.get("./org/img/$str").list()) add(CastleImg(getNextID<CastleImg, CastleImg>(CastleImg::class.java), VImg(vf)))
        }
    }

    class PackCasList(p: UserPack) : CastleList() {
        val pack: UserPack
        override fun getID(): String? {
            return pack.getID()
        }

        override fun toString(): String {
            return pack.desc.name + " (" + size() + ")"
        }

        init {
            pack = p
            map()[pack.desc.id] = this
        }
    }

    override fun <R> getList(cls: Class<*>?, func: Reductor<R, FixIndexMap<*>?>, def: R): R {
        return func.reduce(def, this)
    }

    companion object {
        private const val REG_CASTLE = "castle"
        private const val REG_DEF_CASTLE = "def_castle"
        fun defset(): MutableSet<CastleList> {
            return UserProfile.Companion.getPool<CastleList>(REG_DEF_CASTLE, CastleList::class.java)
        }

        fun from(sta: Stage): Collection<CastleList> {
            val mc: MapColc = sta.map.mc as? PackMapColc ?: return defset()
            val list: MutableList<CastleList> = ArrayList()
            list.addAll(defset())
            val pack: UserPack = (mc as PackMapColc).pack
            list.add(pack.castles)
            for (rel in pack.desc.dependency) list.add(UserProfile.Companion.getUserPack(rel).castles)
            return list
        }

        @ContGetter
        fun getList(str: String): CastleList? {
            return map()[str]
        }

        fun map(): MutableMap<String, CastleList> {
            return UserProfile.Companion.getRegister<CastleList>(REG_CASTLE, CastleList::class.java)
        }
    }
}
